File: src/java/delta/Hash.java
package delta;

import java.math.BigInteger;
import java.util.concurrent.ThreadLocalRandom;

import static delta.Types.*;

/**
 * Karp-Rabin rolling hash (Karp &amp; Rabin 1987; Section 2.1.3).
 *
 * Polynomial fingerprints over the Mersenne prime 2^61-1.
 */
public final class Hash {
    private Hash() {}

    /** Reduce 128-bit value (hi:lo) modulo 2^61-1. */
    public static long modMersenne(long hi, long lo) {
        long P = HASH_MOD;
        long upper = (hi << 3) | (lo >>> 61);
        long lower = lo & P;
        long r = upper + lower;
        if (Long.compareUnsigned(r, P) >= 0) r -= P;
        // Second reduction
        long upper2 = r >>> 61;
        long lower2 = r & P;
        long r2 = upper2 + lower2;
        if (Long.compareUnsigned(r2, P) >= 0) r2 -= P;
        return r2;
    }

    /** (a * b) mod (2^61-1) using 128-bit intermediate. */
    public static long mulmod(long a, long b) {
        long hi = Math.multiplyHigh(a, b);
        long lo = a * b;
        return modMersenne(hi, lo);
    }

    /** Karp-Rabin fingerprint of data[offset..offset+p] (Eq. 1, Section 2.1.3). */
    public static long fingerprint(byte[] data, int offset, int p) {
        long h = 0;
        for (int i = 0; i < p; i++) {
            int b = data[offset + i] & 0xFF;
            long hi = Math.multiplyHigh(h, HASH_BASE);
            long lo = h * HASH_BASE;
            long newLo = lo + b;
            long newHi = hi;
            if (Long.compareUnsigned(newLo, lo) < 0) newHi++;
            h = modMersenne(newHi, newLo);
        }
        return h;
    }

    /** Precompute HASH_BASE^{p-1} mod HASH_MOD. */
    public static long precomputeBp(int p) {
        if (p == 0) return 1;
        long result = 1;
        long base = HASH_BASE;
        int exp = p - 1;
        while (exp > 0) {
            if ((exp & 1) == 1) result = mulmod(result, base);
            base = mulmod(base, base);
            exp >>= 1;
        }
        return result;
    }

    // ── Rolling hash ─────────────────────────────────────────────────

    /** Rolling hash for O(1) incremental fingerprint updates (Eq. 2). */
    public static final class RollingHash {
        private long value;
        private final long bp; // HASH_BASE^{p-1} mod HASH_MOD

        public RollingHash(byte[] data, int offset, int p) {
            this.bp = precomputeBp(p);
            this.value = fingerprint(data, offset, p);
        }

        public long value() { return value; }

        /** Slide window: remove oldByte from left, add newByte to right. */
        public void roll(int oldByte, int newByte) {
            long sub = mulmod(oldByte, bp);
            long v = Long.compareUnsigned(value, sub) >= 0
                ? (value - sub)
                : (HASH_MOD - (sub - value));
            long hi = Math.multiplyHigh(v, HASH_BASE);
            long lo = v * HASH_BASE;
            long newLo = lo + newByte;
            long newHi = hi;
            if (Long.compareUnsigned(newLo, lo) < 0) newHi++;
            value = modMersenne(newHi, newLo);
        }
    }

    // ── Primality testing ────────────────────────────────────────────

    /** Miller-Rabin probabilistic primality test with 100 random witnesses. */
    public static boolean isPrime(long n) {
        return isPrime(n, 100);
    }

    public static boolean isPrime(long n, int k) {
        if (n < 2) return false;
        if (n < 4) return true;
        if (n % 2 == 0) return false;

        BigInteger bn = BigInteger.valueOf(n);
        BigInteger nm1 = bn.subtract(BigInteger.ONE);
        int r = nm1.getLowestSetBit();
        BigInteger d = nm1.shiftRight(r);

        ThreadLocalRandom rng = ThreadLocalRandom.current();
        for (int i = 0; i < k; i++) {
            long a = 2 + rng.nextLong(n - 3);
            BigInteger x = BigInteger.valueOf(a).modPow(d, bn);
            if (x.equals(BigInteger.ONE) || x.equals(nm1)) continue;
            boolean found = false;
            for (int j = 0; j < r - 1; j++) {
                x = x.modPow(BigInteger.TWO, bn);
                if (x.equals(nm1)) { found = true; break; }
            }
            if (!found) return false;
        }
        return true;
    }

    /** Smallest prime >= n. */
    public static long nextPrime(long n) {
        if (n <= 2) return 2;
        if (n % 2 == 0) n++;
        while (!isPrime(n)) n += 2;
        return n;
    }

    // ── CRC-64/XZ (ECMA-182 reflected) — 8-byte output ─────────────
    //
    // Reflected poly: 0xC96C5795D7870F42, Init = XorOut = 0xFFFFFFFFFFFFFFFF.
    // Check value: Crc64.hash8(b"123456789") = 0x995DC9BBDF1939FA big-endian.

    public static final class Crc64 {
        private static final long POLY = 0xC96C5795D7870F42L;
        private static final long[] TABLE;

        static {
            TABLE = new long[256];
            for (int i = 0; i < 256; i++) {
                long c = (long) i;
                for (int j = 0; j < 8; j++)
                    c = (c & 1L) != 0 ? (c >>> 1) ^ POLY : (c >>> 1);
                TABLE[i] = c;
            }
        }

        private Crc64() {}

        /** Compute CRC-64/XZ of data, return 8 bytes big-endian. */
        public static byte[] hash8(byte[] data) {
            long crc = 0xFFFFFFFFFFFFFFFFL;
            for (byte b : data)
                crc = TABLE[(int)((crc ^ b) & 0xFF)] ^ (crc >>> 8);
            crc ^= 0xFFFFFFFFFFFFFFFFL;
            byte[] out = new byte[8];
            for (int i = 0; i < 8; i++)
                out[i] = (byte) (crc >>> (56 - 8 * i));
            return out;
        }
    }
}

File: src/go/delta/hash.go
package delta

import (
	"math/big"
	"math/bits"
)

// ── Karp-Rabin rolling hash (Karp & Rabin 1987; Section 2.1.3) ──
//
// Polynomial fingerprints over the Mersenne prime 2^61-1.

// ModMersenne reduces the 128-bit value (hi:lo) modulo 2^61-1.
func ModMersenne(hi, lo uint64) int64 {
	const P = uint64(HashMod)
	upper := (hi << 3) | (lo >> 61)
	lower := lo & P
	r := upper + lower
	if r >= P {
		r -= P
	}
	// Second reduction.
	upper2 := r >> 61
	lower2 := r & P
	r2 := upper2 + lower2
	if r2 >= P {
		r2 -= P
	}
	return int64(r2)
}

// mulmod computes (a * b) mod (2^61-1) using a 128-bit intermediate.
func mulmod(a, b int64) int64 {
	hi, lo := bits.Mul64(uint64(a), uint64(b))
	return ModMersenne(hi, lo)
}

// Fingerprint computes the Karp-Rabin fingerprint of data[offset..offset+p] (Eq. 1).
func Fingerprint(data []byte, offset, p int) int64 {
	var h int64
	for i := 0; i < p; i++ {
		b := int64(data[offset+i] & 0xFF)
		hi, lo := bits.Mul64(uint64(h), HashBase)
		newLo := lo + uint64(b)
		newHi := hi
		if newLo < lo {
			newHi++
		}
		h = ModMersenne(newHi, newLo)
	}
	return h
}

// PrecomputeBp computes HASH_BASE^{p-1} mod HASH_MOD.
func PrecomputeBp(p int) int64 {
	if p == 0 {
		return 1
	}
	result := int64(1)
	base := int64(HashBase)
	exp := p - 1
	for exp > 0 {
		if exp&1 == 1 {
			result = mulmod(result, base)
		}
		base = mulmod(base, base)
		exp >>= 1
	}
	return result
}

// ── Rolling hash ──

// RollingHash supports O(1) incremental fingerprint updates (Eq. 2).
type RollingHash struct {
	value int64
	bp    int64 // HASH_BASE^{p-1} mod HASH_MOD
}

// NewRollingHash initializes a rolling hash over data[offset..offset+p].
func NewRollingHash(data []byte, offset, p int) *RollingHash {
	return &RollingHash{
		bp:    PrecomputeBp(p),
		value: Fingerprint(data, offset, p),
	}
}

// Value returns the current fingerprint.
func (rh *RollingHash) Value() int64 { return rh.value }

// Roll slides the window: remove oldByte from left, add newByte to right.
func (rh *RollingHash) Roll(oldByte, newByte int) {
	sub := mulmod(int64(oldByte), rh.bp)
	var v int64
	if rh.value >= sub {
		v = rh.value - sub
	} else {
		v = HashMod - (sub - rh.value)
	}
	hi, lo := bits.Mul64(uint64(v), HashBase)
	newLo := lo + uint64(newByte)
	newHi := hi
	if newLo < lo {
		newHi++
	}
	rh.value = ModMersenne(newHi, newLo)
}

// ── Primality testing ──

// Fixed witnesses for deterministic Miller-Rabin.
// Sufficient for all n < 3,317,044,064,679,887,385,961,981 (> 2^81).
// Jaeschke, Math. Comp. 61(204), 1993.
var mrWitnesses = []int64{2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37}

// IsPrime reports whether n is prime using deterministic Miller-Rabin.
func IsPrime(n int64) bool {
	if n < 2 {
		return false
	}
	if n < 4 {
		return true
	}
	if n%2 == 0 {
		return false
	}
	bn := big.NewInt(n)
	nm1 := new(big.Int).Sub(bn, big.NewInt(1))
	r := nm1.TrailingZeroBits()
	d := new(big.Int).Rsh(nm1, r)

	for _, a := range mrWitnesses {
		if a >= n {
			break
		}
		x := new(big.Int).Exp(big.NewInt(a), d, bn)
		if x.Cmp(big.NewInt(1)) == 0 || x.Cmp(nm1) == 0 {
			continue
		}
		found := false
		for j := uint(0); j < r-1; j++ {
			x.Exp(x, big.NewInt(2), bn)
			if x.Cmp(nm1) == 0 {
				found = true
				break
			}
		}
		if !found {
			return false
		}
	}
	return true
}

// NextPrime returns the smallest prime >= n.
func NextPrime(n int64) int64 {
	if n <= 2 {
		return 2
	}
	if n%2 == 0 {
		n++
	}
	for !IsPrime(n) {
		n += 2
	}
	return n
}
